src/runtime/current_thread/runtime.rs

```diff
@@ -0,0 +1,111 @@
+        // We need a reactor to receive events about IO objects from kernel
+        let reactor = Reactor::new()?;
+        let reactor_handle = reactor.handle();
+
+        // Place a timer wheel on top of the reactor. If there are no timeouts to fire, it'll let the
+        // reactor pick up new some new external events.
+        let timer = Timer::new(reactor);
+        let timer_handle = timer.handle();
+
+        // And now put a single-threaded executor on top of the timer. When there are no futures ready
+        // to do something, it'll let the timer or the reactor to generate some new stimuli for the
+        // futures to continue in their life.
+        let executor = CurrentThread::new_with_park(timer);
+
+        let runtime = Runtime { reactor_handle, timer_handle, executor };
+        Ok(runtime)
+    }
+
+    /// Blocks and runs the given future.
+    ///
+    /// This is similar to running a runtime, but uses only the current thread.
+    pub fn block_on<F: Future<Item = (), Error = ()>>(&mut self, f: F)
+            -> Result<(), current_thread::BlockError<()>> {
+        let Runtime { ref reactor_handle, ref timer_handle, ref mut executor } = *self;
+
+        // Binds an executor to this thread
+        let mut enter = tokio_executor::enter().expect("Multiple executors at once");
+
+        // This will set the default handle and timer to use inside the closure and run the future.
+        tokio_reactor::with_default(&reactor_handle, &mut enter, |enter| {
+            timer::with_default(&timer_handle, enter, |enter| {
+                // The TaskExecutor is a fake executor that looks into the current single-threaded
+                // executor when used. This is a trick, because we need two mutable references to the
+                // executor (one to run the provided future, another to install as the default one). We
+                // use the fake one here as the default one.
+                let mut default_executor = current_thread::TaskExecutor::current();
+                tokio_executor::with_default(&mut default_executor, enter, |enter| {
+                    let mut executor = executor.enter(enter);
+                    // Run the provided future
+                    executor.block_on(f)
+                })
+            })
+        })
+    }
+
+    /// Blocks and runs all the other futures that are still left in the executor
+    ///
+    /// This is similar to running a runtime, but uses only the current thread.
+    pub fn run(&mut self) -> Result<(), current_thread::RunError> {
+        let Runtime { ref reactor_handle, ref timer_handle, ref mut executor } = *self;
+
+        // Binds an executor to this thread
+        let mut enter = tokio_executor::enter().expect("Multiple executors at once");
+
+        // This will set the default handle and timer to use inside the closure and run the future.
+        tokio_reactor::with_default(&reactor_handle, &mut enter, |enter| {
```

Comment:
It looks like most of the body of the function is duplicated from above. It would be nice if this could be extracted.